### PR TITLE
[pkg/stanza/recombine] fix: recombine aggregation problem fix

### PR DESCRIPTION
**Pull request by yutingcaicyt**
**Link to tracking Issue:** #21241

### CODE REVIEW

File: pkg/stanza/operator/transformer/recombine/recombine_test.go [hunk 1]
```diff
@@ -511,3 +556,39 @@ func TestTimeoutWhenAggregationKeepHappen(t *testing.T) {
 	}
 	require.NoError(t, recombine.Stop())
 }
+
+func TestSourceBatchDelete(t *testing.T) {
```
Comment (djaglowski) [inline pkg/stanza/operator/transformer/recombine/recombine_test.go:560]: Would you mind adding a test that triggers based on max batch size, and validates the logs that are emitted by the operator. (similar to your original example of 100 lines emitted as 50 + 50).

**Review by djaglowski — COMMENTED** [review]
Thanks for taking this on @yutingcaicyt. Generally looks good to me but it's quite nuanced so I have a couple requests to ensure we do not regress in the future.

File: pkg/stanza/operator/transformer/recombine/recombine.go [hunk 2]
```diff
@@ -290,6 +290,9 @@ func (r *Transformer) addToBatch(_ context.Context, e *entry.Entry, source strin
 			return
 		}
 	} else {
+		if len(batch.entries) == 0 {
+			batch.firstEntryObservedTime = e.ObservedTimestamp
```
[the remark this answers is not in the document]
Comment (yutingcaicyt) [reply pkg/stanza/operator/transformer/recombine/recombine.go:296]: comment added

File: pkg/stanza/operator/transformer/recombine/recombine_test.go (hunk 1 above)
Comment (yutingcaicyt) [reply pkg/stanza/operator/transformer/recombine/recombine_test.go:587]: yeah, a new test added.

**Review by djaglowski — APPROVED** [review]

**Pull request merged**